src/std/misc/text.ss

```diff
@@ -13,3 +19,147 @@
      (let (path (gx#core-resolve-path #'path (stx-source stx)))
        (with-syntax ((txt (call-with-input-file path read-all-as-string)))
          #'(quote txt))))))
+
+;; quasistring utility functions
+(begin-syntax
+  (def qs-err-missing-close "Malformed quasistring: closing } required after s-exp")
+
+  (def (stx-format stx)
+    (if (stx-string? stx) ;; don't format strings
+      stx
+      (with-syntax ((stx stx))
+        #'(format "~a" stx))))
+
+  (def (map-format stx)
+    (stx-map stx-format stx))
+
+  ;; custom eval to capture the proper variables
+  (def (eval** stx ctx)
+    (parameterize ((current-expander-phi (1+ (current-expander-phi))))
+      (let (stx (datum->syntax ctx stx))
+        (eval-syntax stx #t))))
+
+  (def (parse-quasistring stx ctx rt-symbol-proc output-proc (expansion-time-eval #f))
+    ;; rt-symbol-proc wraps sexp if sexp is a run time symbol
+    ;; output-proc is what we return from the macro
+    ;; if expansion-time-eval is #t, allow for ##{} template variables
+    (def (lp (port (open-input-string (stx-e stx)))
+             (c    (read-char port))
+             (ops  [])  ;; arguments of string-append
+             (str  "")) ;; container to concatenate strings
+      (cond
+       ((eq? c #\#)
+        (let (c+1 (read-char port))
+          (cond
+           ((eq? c+1 #\{) ;; run time template variables
+            (let ((sexp (read port))
+                  (c+2  (read-char port)))
+              (if (eq? c+2 #\})
+                (if (symbol? sexp)
+                  (lp port (read-char port) (cons* (rt-symbol-proc sexp) str ops) "")
+                  (lp port (read-char port) (cons* (datum->syntax ctx sexp) str ops) ""))
+                (lp port (read-char port) ops (string-append str (list->string [c c+1 c+2]))))))
+           ((eq? c+1 #\#) ;; expansion time template variables
+            (if expansion-time-eval
+              (let (c+2 (read-char port))
+                (if (eq? c+2 #\{)
+                  (let* ((sexp     (read port))
+                         (sexp-str (format "~a" (eval** sexp ctx)))
+                         (c+3      (read-char port)))
+                    (if (eq? c+3 #\})
+                      ;; allow for nested templates
+                      (let* ((rest (read-all-as-string port))
```

Comment:
Here we must recurse since we might have expanded into a sub-template. Thus we:

1. Read the expansion-time template variable
2. Expand and format it to a string
3. Prepend it to the rest of our current quasistring
4. Start the parsing from the newly concatenated quasistring, which might or might not contain other sub-templates.